app_dart/lib/src/request_handlers/refresh_chromebot_status.dart

```diff
@@ -73,25 +71,38 @@ class RefreshChromebotStatus extends ApiRequestHandler<Body> {
 
   /// Update chromebot tasks statuses in datastore for [builder],
   /// based on latest [luciTasks] statuses.
-  Future<void> _updateStatus(
-      LuciBuilder builder,
-      List<String> branches,
-      DatastoreService datastore,
-      Map<LuciBuilder, List<LuciTask>> luciTasks) async {
-    for (String branch in branches) {
-      await for (FullTask task in datastore.queryRecentTasks(
-          taskName: builder.taskName, branch: branch)) {
-        for (LuciTask luciTask in luciTasks[builder]) {
-          if (luciTask.commitSha == task.commit.sha &&
-              luciTask.ref == 'refs/heads/$branch') {
-            final Task update = task.task;
-            update.status = luciTask.status;
-            await datastore.insert(<Task>[update]);
-            // Stop updating task whenever we find the latest status of the same commit.
-            break;
-          }
+  Future<void> _updateStatus(LuciBuilder builder, String branch,
+      DatastoreService datastore, Map<String, List<LuciTask>> luciTasks) async {
+    final List<FullTask> datastoreTasks = await datastore
+        .queryRecentTasks(taskName: builder.taskName, branch: branch)
+        .toList();
+    final Set<LuciTask> updatedLuciTasks = <LuciTask>{};
+
+    /// Scan both [datastoreTasks] and [luciTasks] from old to new, since [luciTasks] may
+    /// contain new re-run builds which are not in datastore yet. If matched, then
+    /// update accordingly.
+    for (FullTask datastoreTask in datastoreTasks.reversed) {
+      if (!luciTasks.containsKey(datastoreTask.commit.sha)) {
+        continue;
+      }
+      for (LuciTask luciTask in luciTasks[datastoreTask.commit.sha].reversed) {
+        if (!updatedLuciTasks.contains(luciTask) &&
+            _buildNumberMatched(datastoreTask, luciTask)) {
```

Comment:
Nit: I think the `!updatedLuciTasks.contains(luciTask)` check is unnecessary, since the inner loop breaks right after it finds the first match.

Reply:
It is needed for cases where there are reruns (this is exactly why this PR) for a commit.

Reply:
A `datastoreTask` is expected to be updated by no more than one `luciTask`. 
- When `datastoreTask.task.buildNumber == null`, it expects the newest `luciTask`
- When not null, it expects the newest `luciTask` with matching `buildNumber`.

Therefore we can loop the luciTask from new to old and break when a first match is found.

Reply:
A `datastoreTask` is expected to be updated by no more than one `luciTask`. 
- When `datastoreTask.task.buildNumber == null`, it expects the newest `luciTask`
- When not null, it expects the newest `luciTask` with matching `buildNumber`.

Therefore we can loop the luciTask from new to old and break when a first match is found.

Reply:
When new builds or re-runs happen, they will be inserted to datastore immediately, but may not scheduled yet in luci. For such case, there will not be one-to-one mapping between `datastoreTask` and `luciTasks` from new to old. That's why I changed the looping order from old to new. 

The old strategy only needs to keep the latest status of a commit, so it works as expected when looping from new to old.

Reply:
Very nit: OK then this apparently can be simplified to linear time similar to merging two sorted arrays. And the comments need more clarification for future readers.

Reply:
The logic is already "almost" in linear time complexity, thanks to the "complex" map we created. When iterating through `datastoreTask`, we compare directly with `luciTasks[datastoreTask.commit.sha]`. For commits without reruns, this is just one time comparison. 

I have updated the doc. to make it clearer.